internal/kafka/internal/config/kafka_supported_instance_types.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	return nil
+}
+
+type SupportedKafkaInstanceTypesConfig struct {
+	SupportedKafkaInstanceTypes []KafkaInstanceType `yaml:"supported_instance_types"`
+}
+
+func (s *SupportedKafkaInstanceTypesConfig) validate() error {
+	existingInstanceTypes := make(map[string]int, len(s.SupportedKafkaInstanceTypes))
+
+	for _, KafkaInstanceType := range s.SupportedKafkaInstanceTypes {
+		if _, ok := existingInstanceTypes[KafkaInstanceType.Id]; ok {
+			return fmt.Errorf("Kafka instance type id '%s' was defined more than once.", KafkaInstanceType.Id)
+		}
+		existingInstanceTypes[KafkaInstanceType.Id]++
+
+		if err := KafkaInstanceType.validate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type KafkaSupportedInstanceTypesConfig struct {
+	Configuration     SupportedKafkaInstanceTypesConfig
+	ConfigurationFile string
+}
+
+func NewKafkaSupportedInstanceTypesConfig() *KafkaSupportedInstanceTypesConfig {
+	return &KafkaSupportedInstanceTypesConfig{
+		ConfigurationFile: "config/kafka-instance-types-configuration.yaml",
+	}
+}
+
+func (s *SupportedKafkaInstanceTypesConfig) GetKafkaInstanceTypeByID(instanceType string) (*KafkaInstanceType, error) {
+	for _, t := range s.SupportedKafkaInstanceTypes {
+		if t.Id == instanceType {
+			ret := t
+			return &ret, nil
+		}
+	}
+	return nil, fmt.Errorf("Unable to find kafka instance type for '%s'", instanceType)
+}
+
+func (kp *KafkaInstanceType) GetKafkaInstanceSizeByID(sizeId string) (*KafkaInstanceSize, error) {
+	for _, size := range kp.Sizes {
+		if size.Id == sizeId {
+			ret := size
+			return &ret, nil
+		}
+	}
+	return nil, fmt.Errorf("Kafka instance size id: '%s' not found for '%s' instance type", sizeId, kp.Id)
+}
+
+type Plan string
+
+func (p Plan) String() string {
+	return string(p)
+}
+
+const numComponentsInPlanFormat = 2
+
+func (p Plan) GetInstanceType() (string, error) {
+	t := strings.Split(strings.TrimSpace(p.String()), ".")
+	if len(t) != numComponentsInPlanFormat {
+		return "", errors.New(errors.ErrorGeneral, fmt.Sprintf("Unsupported plan provided: '%s'", p))
+	}
+	return t[0], nil
+}
+
+func (p Plan) GetSizeID() (string, error) {
+	t := strings.Split(strings.TrimSpace(p.String()), ".")
+	if len(t) != numComponentsInPlanFormat {
+		return "", errors.New(errors.ErrorGeneral, fmt.Sprintf("Unsupported plan provided: '%s'", p))
+	}
+	return t[1], nil
+}
+
+type Quantity string
+
+func (q *Quantity) String() string {
+	return string(*q)
+}
+
+func (q *Quantity) ToFloat32() (float32, error) {
+	if p, err := resource.ParseQuantity(string(*q)); err != nil {
+		return 0, err
+	} else {
+		return float32(p.Value()), nil
+	}
+}
+
+func (q *Quantity) ToK8Quantity() (*resource.Quantity, error) {
+	if p, err := resource.ParseQuantity(string(*q)); err != nil {
+		return nil, err
+	} else {
+		return &p, nil
+	}
+}
+
+func (q *Quantity) IsEmpty() bool {
+	return q == nil
+}
```

Comment:
This file is defining the following objects:

- KafkaInstanceType
- SupportedKafkaInstanceTypesConfig
- KafkaSupportedInstanceTypesConfig
- Plan
- Quantity

Together with their methods. IMO, this makes the code hard do read/maintains.
Wouldn't be better to place each object in its own file?

Reply:
It's a good suggestion @ziccardi , but can be done in a separate PR.

Reply:
:+1